File: evals/refs/det.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
)

func main() {
	in := bufio.NewScanner(os.Stdin)
	in.Buffer(make([]byte, 1<<20), 1<<20)
	in.Split(bufio.ScanWords)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()
	read := func() string { in.Scan(); return in.Text() }
	T, _ := strconv.Atoi(read())
	for c := 0; c < T; c++ {
		N, _ := strconv.Atoi(read())
		M := make([][]float64, N)
		for i := 0; i < N; i++ {
			M[i] = make([]float64, N)
			for j := 0; j < N; j++ {
				v, _ := strconv.ParseFloat(read(), 64)
				M[i][j] = v
			}
		}
		sign := 1.0
		zero := false
		for col := 0; col < N; col++ {
			pivot := col
			for r := col + 1; r < N; r++ {
				if math.Abs(M[r][col]) > math.Abs(M[pivot][col]) {
					pivot = r
				}
			}
			if pivot != col {
				M[col], M[pivot] = M[pivot], M[col]
				sign = -sign
			}
			if math.Abs(M[col][col]) < 1e-15 {
				zero = true
				break
			}
			for r := col + 1; r < N; r++ {
				factor := M[r][col] / M[col][col]
				for c2 := col; c2 < N; c2++ {
					M[r][c2] -= factor * M[col][c2]
				}
			}
		}
		var det float64
		if zero {
			det = 0
		} else {
			det = sign
			for i := 0; i < N; i++ {
				det *= M[i][i]
			}
		}
		fmt.Fprintf(out, "%.6f\n", det)
	}
}

File: evals/refs/det.py
import sys
data = sys.stdin.read().split()
idx = 0
T = int(data[idx]); idx += 1
out = []
for _ in range(T):
    N = int(data[idx]); idx += 1
    M = []
    for r in range(N):
        row = [float(data[idx+c]) for c in range(N)]
        M.append(row); idx += N
    sign = 1.0
    for col in range(N):
        # Partial pivot
        pivot = col
        for r in range(col+1, N):
            if abs(M[r][col]) > abs(M[pivot][col]):
                pivot = r
        if pivot != col:
            M[col], M[pivot] = M[pivot], M[col]
            sign = -sign
        if abs(M[col][col]) < 1e-15:
            sign = 0
            break
        for r in range(col+1, N):
            factor = M[r][col] / M[col][col]
            for c2 in range(col, N):
                M[r][c2] -= factor * M[col][c2]
    if sign == 0:
        det = 0.0
    else:
        det = sign
        for i in range(N):
            det *= M[i][i]
    out.append(f'{det:.6f}')
sys.stdout.write('\n'.join(out) + '\n')
